015-kafka-topic-encryption.md

```diff
@@ -0,0 +1,169 @@
+
+For each topic to be encrypted, a policy exists detailing:
+
+- topic name
+- encryption algorithm(s), cipher suite, and encryption/decryption steps
+- key management server address
+- key identifier
+- optional initialization information
+
+In our reference implementation, policies are encoded in JSON. An example topic encryption policy,
+used in testing, is shown below:
+
+```
+[
+ {
+   "topic" : "juicy-topic",
+   "kms_url" : "https://kms.abc.com",
+   "kms_instance": "11111111-2222-3333-4444-55555555555",
+   "key_ref" : "e2a73a6c4bf9",
+   "credential" : "dkT4WSCnrlNE7"
+ },
+ {
+   "topic" : "creditcard-data",
+   "kms_url" : "https://kms.abc.com",
+   "kms_instance": "11111111-2222-3333-4444-55555555555",
+   "key_ref" : "9957cdf5f32a",
+   "credential" : "dkT4WSCnrlNE7"
+ }
+]
+```
+
+In this example, two topics will be encrypted with two different keys residing in the same KMS.
+
+### Message Handling
+Message handling is concerned with facilities for intercepting messages, deserializing and inspecting them, invoking encryption as accorded by policy, modifying messages, and finally serializing and forwarding messages onward to Kafka brokers and clients. To accomplish encryption and decryption of data payloads, only Kafka *Produce requests* and *Fetch responses* need to be examined and potentially modified. High quality open source libraries exist already for implementing message introspection and serialization/deseriaization, e.g.,
+https://github.com/Shopify/sarama.
+
+### Message Encryption
+As described in [Kafka message format documentation](https://kafka.apache.org/documentation/#messageformat), data is transmitted in [record batches](https://kafka.apache.org/documentation/#recordbatch). Within a batch, each [record](https://kafka.apache.org/documentation/#record) is comprised of several fields of which two may contain data and therefore are of direct relevance for topic encryption: the _value_ field and the _headers_ field, an array of optional [record header structures](https://kafka.apache.org/documentation/#recordheader). _Key_ fields are never encrypted in this proposal. Any functionality relying on keys such as partitioning and compaction are unaffected and thus preserved by topic encryption.
+
+An example is now described to illuminate the essential details of the message encryption process.
+The encryption module is embedded in a proxy and the proxy, as an intermediary, terminates  connections between clients and the broker, placing the proxy in the position of intercepting all Kafka traffic. The message handling component examines all messages to identify the apikey (i.e., the Kafka message type). All message types besides Produce and Fetch are  immediately forwarded without any processing. With the help of the message handling library, Produce requests are deserialized and inspected more deeply to identify the topic for which records are destined.
```

Comment:
> All message types besides Produce and Fetch are immediately forwarded without any processing.

You probably want to do something with `ApiVersions` requests. They're used to inform the client about the protocol versions (for each api key) supported by the broker. So it would be possible for a new broker version to add a new version of a `Produce` request which the proxy didn't understand, and by just sending the `ApiVersions` response to the client unchanged a client which also supported that version of `Produce` would use it. So for the `Produce` and `Fetch` APIs the max version in the `ApiVersions` response should the `min(brokerMaxVersion, proxyMaxVersion)`, likewise the min versions should be the `max(brokerMinVersion, proxyMinVersion)`. If you're using Envoy it's possible that it already handles this, but it's worth checking.

Reply:
Agree, good point. Will add to the next update.